From 9d58f420849912f4ae00ecfc9f308835aa65b57e Mon Sep 17 00:00:00 2001
From: Jan Stancek <jstancek@redhat.com>
Date: Mon, 14 Jul 2014 14:02:09 -0400
Subject: [fs] fanotify/compat: fix sys_fanotify_mark

Message-id: <5523a2ee06f671c31c772baa91050472a28c4ae5.1405346097.git.jstancek@redhat.com>
Patchwork-id: 85790
O-Subject: [RHEL7.1 PATCH] compat: fix sys_fanotify_mark
Bugzilla: 1119318
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Alexander Viro <aviro@redhat.com>

Bugzilla: 1119318
Upstream Status: linux.git commit 592f6b842f64e416c7598a1b97c649b34241e22d
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7696534
Tested: with fanotify testcases from LTP

commit 592f6b842f64e416c7598a1b97c649b34241e22d
Author: Heiko Carstens <heiko.carstens@de.ibm.com>
Date:   Mon Jan 27 17:07:19 2014 -0800
    compat: fix sys_fanotify_mark

    Commit 91c2e0bcae72 ("unify compat fanotify_mark(2), switch to
    COMPAT_SYSCALL_DEFINE") added a new unified compat fanotify_mark syscall
    to be used by all architectures.

    Unfortunately the unified version merges the split mask parameter in a
    wrong way: the lower and higher word got swapped.

    This was discovered with glibc's tst-fanotify test case.

    Signed-off-by: Heiko Carstens <heiko.carstens@de.ibm.com>
    Reported-by: Andreas Krebbel <krebbel@linux.vnet.ibm.com>
    Cc: "James E.J. Bottomley" <jejb@parisc-linux.org>
    Acked-by: "David S. Miller" <davem@davemloft.net>
    Acked-by: Al Viro <viro@ZenIV.linux.org.uk>
    Cc: Benjamin Herrenschmidt <benh@kernel.crashing.org>
    Cc: Ingo Molnar <mingo@redhat.com>
    Cc: Ralf Baechle <ralf@linux-mips.org>
    Cc: <stable@vger.kernel.org>        [3.10+]
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Jan Stancek <jstancek@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/notify/fanotify/fanotify_user.c b/fs/notify/fanotify/fanotify_user.c
index 77cc85d..f1680cd 100644
--- a/fs/notify/fanotify/fanotify_user.c
+++ b/fs/notify/fanotify/fanotify_user.c
@@ -867,9 +867,9 @@ COMPAT_SYSCALL_DEFINE6(fanotify_mark,
 {
  return sys_fanotify_mark(fanotify_fd, flags,
 #ifdef __BIG_ENDIAN
-    ((__u64)mask1 << 32) | mask0,
-#else
     ((__u64)mask0 << 32) | mask1,
+#else
+    ((__u64)mask1 << 32) | mask0,
 #endif
      dfd, pathname);
 }
-- 
1.7.1